Update web components snippets (JS & TS)

kylegach opened the pull request:
## What I did

- Update the web-components snippets (both JS & TS) for these pages:
  - /docs/get-started/whats-a-story
  - /docs/get-started/setup
  - /docs/writing-stories/introduction
  - /docs/writing-stories/args (WIP)
  - /docs/writing-stories/parameters
  - /docs/writing-stories/decorators
  - /docs/writing-stories/play-function
  - /docs/writing-stories/loaders
  - /docs/writing-stories/naming-components-and-hierarchy
  - /docs/writing-stories/build-pages-in-storybook
  - /docs/writing-stories/stories-for-multiple-components
  - /docs/api/csf
- Convert some snippets to "common" ones
- Remove unused snippets
<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `master` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

Maintainers: Please tag your pull request with at least one of the following:
`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

-->

kylegach commented on docs/snippets/web-components/button-group-story.js.mdx:
```diff
@@ -12,16 +12,15 @@ export default {
+   * See https://storybook.js.org/docs/7.0/web-components/api/csf
+   * to learn how to use render functions.
+   */
+  render: (args) => ButtonGroup(args),
```
It seems like we should move the `render` function into the Meta whenever it makes sense to do so, as that will match the majority of the users' experience. Agreed?

kylegach replied:
Note that I have _not_ applied this change to all of the files modified/created here, until we agree on the way forward.

jonniebigodes replied:
Yes, we should, as it's the default experience when a user adds Storybook to their Web Component's project. I've already been applying it in my pull request.

kylegach commented on docs/snippets/web-components/button-story-component-decorator.ts.mdx:
```diff
@@ -0,0 +1,29 @@
+};
+export default meta;
+
+type Story = StoryObj;
```
When the component doesn't have a prop type available (e.g. `import './demo-button'` vs. `import { Button, ButtonProps } from './demo-button'`) then there's nothing to pass to `StoryObj` and this assignment is a bit silly. But it's also more consistent across situations. Wdyt?

jonniebigodes replied:
I know, and I had the same question as you did, and I went for consistency and left in the StoryObj type as an additional precaution to prevent unforeseen questions about missing typings.

kylegach replied:
To be clear, the alternatives are:

A:
```js
type Story = StoryObj;

export const Example: Story = { ... }
```

or

B:
```js
export const Example: StoryObj = { ... }
```

A is more consistent with the more prevalent case of `type Story = StoryObj<ComponentProps>`, but without passing a generic, the assignment could be seen as a bit silly or unnecessary. Hence, B.

Looking at your PR, it looks like you might've gone with B?

jonniebigodes replied:
To some degree, yes I went with that, but it's a bit situational and something we'll need to consider moving forward as we build more accurate examples

kylegach commented on docs/snippets/web-components/login-form-with-play-function.js.mdx:
````diff
@@ -0,0 +1,44 @@
+```js
````
This was marked `n/a` on the [core snippet audit table](https://www.notion.so/chromatic-ui/Storybook-core-framework-snippet-audit-eb7bbbcc8e9a480f81fb5eaacc4f5c49), but I wasn't sure why. It's true that play functions can run into problems when web components use the Shadow DOM, but I don't think we need to assume that will be the case and omit the example entirely.

jonniebigodes replied:
I'm good with it. I'll update the snippet database accordingly and remove the example altogether. Once and if [this](https://github.com/konnorrogers/shadow-dom-testing-library) is added into the testing library and we make the change on our end, we can revisit this and add it back as with other examples. Sounds good?

kylegach replied:
I think you misunderstood my comment.

I'm suggesting to _add_ this web components snippet. Because, even with the Shadow DOM complications, it still can be useful for other cases.

jonniebigodes replied:
Sorry for the misunderstanding. We can add it, sure, and while we're at it, add the remaining ones if you're ok with it.

jonniebigodes commented on docs/snippets/web-components/button-group-story.ts.mdx:
````diff
@@ -0,0 +1,33 @@
+```ts
+// demo-button-group.stories.ts
+import type { Meta, StoryObj } from '@storybook/web-components';
+
+import { ButtonGroup, ButtonGroupProps } from './demo-button-group';
+
+//👇 Imports the Button stories
+import * as ButtonStories from './demo-button.stories';
````
@kylegach, when you're able, can you update the imports across your pr to the following structure so it matches the rest of the snippets and examples we provide once Storybook is added to a Web Components' project:

```ts
// ButtonGroup.stories.ts

import type { Meta, StoryObj } from '@storybook/web-components';

import type { ButtonGroupProps } from './ButtonGroup';
import { ButtonGroup } from './ButtonGroup';

const meta: Meta<ButtonGroupProps> = {
  title: 'ButtonGroup',
  render: (args) => ButtonGroup(args),
};

// Remainder implementation
```

The main reasons for this are that as of the latest changes, we're not defining the examples as `demo-*` but instead calling them by their actual names. Also, removing the comment regarding the title was intentional, as, during my testing, if we don't name the story, Storybook throws an error.

jonniebigodes reviewed (changes requested):
@kylegach, thanks for the assist on this. I've left one small item to address, and it seems we're on the same page regarding how to document the examples.

kylegach commented on docs/snippets/web-components/register-component-with-play-function.js.mdx:
```diff
@@ -0,0 +1,41 @@
+
+export default {
+  title: 'RegistrationForm',
+  component: RegistrationForm,
```
According to https://github.com/storybookjs/storybook/blob/next/code/renderers/web-components/src/render.ts#L17-L31 (as well as https://storybook.js.org/docs/7.0/web-components/api/csf#upgrading-from-csf-2-to-csf-3), the `web-components` renderer has a default/implicit render function. Which means all of these play function snippets (and probably many more) can be made "common" and use the approach used here: https://storybook.js.org/docs/7.0/react/writing-stories/parameters#story-parameters. Wdyt?

jonniebigodes replied:
Sounds good to me. If we could make it more platform agnostic as possible, we should. Less maintenance cost and less error-prone issues

kylegach committed:
Update web components snippets (JS & TS)

- For these pages (so far):
  - /docs/get-started/whats-a-story
  - /docs/get-started/setup
  - /docs/writing-stories/introduction
  - /docs/writing-stories/args
  - /docs/writing-stories/parameters
  - /docs/writing-stories/decorators
  - /docs/writing-stories/play-function
  - /docs/writing-stories/loaders
  - /docs/writing-stories/build-pages-in-storybook
  - /docs/writing-stories/stories-for-multiple-components
  - /docs/api/csf

- Convert some snippets to "common" ones
- Remove unused snippets

kylegach commented on docs/snippets/web-components/document-screen-fetch.js.mdx:
````diff
@@ -0,0 +1,65 @@
+```js
````
@kasperpeulen — Does this implementation look right to you?

Used here: https://storybook.js.org/docs/7.0/web-components/writing-stories/build-pages-with-storybook#mocking-rest-requests-with-msw-addon

kasperpeulen replied:
Looks good!

kylegach commented on docs/snippets/web-components/simple-page-implementation.js.mdx:
````diff
@@ -0,0 +1,28 @@
+```js
````
@kasperpeulen — Does this implementation look right to you?

Used here: https://storybook.js.org/docs/7.0/web-components/writing-stories/build-pages-with-storybook#args-composition-for-presentational-screens

kasperpeulen replied:
looks good as well

kylegach commented on docs/snippets/web-components/simple-page-implementation.ts.mdx:
````diff
@@ -0,0 +1,39 @@
+```ts
````
@kasperpeulen — Does this implementation look right to you?

Used here: https://storybook.js.org/docs/7.0/web-components/writing-stories/build-pages-with-storybook#args-composition-for-presentational-screens

jonniebigodes reviewed (changes requested):
@kylegach this is looking great. Just a couple of small tweaks from my end and this should be good to go.

kylegach committed:
Address feedback

- Newline following filename comment
- Fix URLs in comments
- Ensure consistent usage of `ts` vs `tsx` language

jonniebigodes commented on docs/writing-stories/loaders.md:
```diff
@@ -22,6 +22,7 @@ Loaders are helpful when you need to load story data externally (e.g., from a re
     'vue/loader-story.3.js.mdx',
     'vue/loader-story.ts-3.ts.mdx',
     'angular/loader-story.ts.mdx',
+    'web-components/loader-story.ts.mdx',
```
@kylegach, sorry for missing this earlier and correct me if I'm wrong, but you're referencing the `web-components/loader-story.ts.mdx`, but the file you've added is `web-components/loader-story.js.mdx`. Shouldn't it be mentioned as such and probably we'll need to add the TS variant as well? Don't you think?

kylegach replied:
Good catch! 🙏

jonniebigodes reviewed (changes requested):
It seems to be just one small item to address that was about to slip under the radar. Let me know once you've addressed it so that we finally merge this.

kasperpeulen commented on docs/snippets/web-components/simple-page-implementation.ts.mdx:
```diff
@@ -0,0 +1,39 @@
+    user: Record<string, unknown>;
+    document: Record<string, unknown>;
+    subdocuments: Array<Record<string, unknown>>;
+  } = {}
```
I can remember issues with assigning default values in web components, but that may not be true with decorators?

kylegach replied:
Docs here: https://lit.dev/docs/api/decorators/#property

kylegach committed:
Address comments

- Add missing snippet

kylegach committed:
Format snippets

kylegach commented:
The 2 Chromatic diffs are almost certainly unrelated to this PR, which only touches docs & snippets.

jonniebigodes reviewed (approved):
Looks good to me. I was about to mention the formatting, but you already updated it with a recent commit. Thanks for that. Merge when able @kylegach
